File: examples/gaia.py
#!/usr/bin/env python3
"""gaia — the world-memory ACT on the actor-mesh.

GAIA is the engine's central world-memory module as a live actor: eight
layers (time, weather, entropy, gravity, wind, temp, light, memory) fold
out of ONE pseudorandom seed, deterministically, forever. The actor
publishes every layer to `gaia.state` on the mesh's compact wire
(single-char keys) — the field the NPC actors walk on, the clock the
keeper folds. Same seed + same tick ⇒ same universe; nothing is stored,
everything is re-derivable.

Run with uv (needs a nats-server on 4222):
    uv run --with nats-py python examples/gaia.py
    uv run --with nats-py python examples/gaia.py --brief "sanctuary·overworld" --tick 3600
"""

import argparse
import asyncio
import hashlib
import json
import math
import os

import nats

HOST = os.environ.get("NATS_URL", "nats://127.0.0.1:4222")

WEATHER = ["calm", "mist", "wind", "rain", "storm", "clear", "overcast", "frost"]


def _h(base: int, epoch: int) -> int:
    h = (base ^ (epoch * 2654435761)) & 0xFFFFFFFF
    h = ((h ^ (h >> 16)) * 2246822519) & 0xFFFFFFFF
    h = ((h ^ (h >> 13)) * 3266489917) & 0xFFFFFFFF
    return (h ^ (h >> 16)) & 0xFFFFFFFF


def _r2(x: float) -> float:
    # JS-style Math.round(x*100)/100 — half-up, to mirror gaia.ts exactly
    return int(x * 100 + 0.5) / 100


class Gaia:
    """The eight layers of the world, folded from one seed.

    Mirrors quantTernEngine/gaia.ts field for field: the same brief and
    tick give the same universe in Node and in Python — the world core is
    portable, the wire is the same.
    """

    def __init__(self, seed_text: str, tick: int = 0):
        self.brief = seed_text
        self.seed = seed_text  # the brief IS the seed (words → universe)
        self.base = self._seed_base(seed_text)
        self.t = max(0, tick)

    @staticmethod
    def _seed_base(text: str) -> int:
        # sha256('gaia·' + text)[:8] — the same root number as gaia.ts
        d = hashlib.sha256(f"gaia·{text}".encode()).hexdigest()[:8]
        return int(d, 16) & 0xFFFFFFFF

    def state(self):
        base, t = self.base, self.t
        w0 = _h(base, t // 60) % 360
        w1 = _h(base, t // 60 + 1) % 360
        f = (t % 60) / 60
        direction = int(w0 + ((w1 - w0 + 360) % 360) * f + 0.5) % 360
        speed = 1 + (_h(base, t // 60 + 7) % 12)
        gust = _r2(speed + ((_h(base, t) % 100) / 100) * 4)
        temp = _r2(
            15
            + 8 * math.sin(t / 864000 * math.pi * 2)
            + 4 * math.sin(t / 86400 * math.pi * 2)
            + ((_h(base, t % 3600) % 100) / 100 - 0.5) * 2,
        )
        return {
            "t": t,
            "w": WEATHER[_h(base, t // 300) % len(WEATHER)],
            "e": _r2(t / 86400 + (_h(base, 0) % 1000) / 1000),
            "g": _r2(9.7 + (_h(base, 1) % 100) / 100),
            "v": [direction, speed, gust],
            "p": temp,
            "l": _r2(max(0.0, math.sin(t / 86400 * math.pi * 2))),
            "m": self._memory(base, t),
        }

    @staticmethod
    def _memory(base: int, t: int) -> str:
        mem = _h(base, 0)
        for i in range(1, min(t // 3600, 4096) + 1):
            mem = _h(mem, i)
        return f"{mem:08x}"

    def step(self, dt: int = 1):
        self.t += dt
        return self.state()


async def run(brief: str, tick: int, rate: float):
    nc = await nats.connect(HOST)
    gaia = Gaia(brief, tick)
    print(f"⟦ GAIA · the world runs without you · {brief} · t={gaia.t} ⟧")
    while True:
        s = gaia.step()
        body = json.dumps(s).encode()
        try:
            await nc.publish("gaia.state", body)
        except Exception:
            pass
        print(f"  t{s['t']:>6}  {s['w']:8}  e {s['e']:.2f}  g {s['g']:.2f}  "
              f"v {s['v'][0]}°@{s['v'][1]}m/s  p {s['p']:.1f}°  l {s['l']:.2f}  m {s['m']}")
        await asyncio.sleep(rate)


def main():
    p = argparse.ArgumentParser()
    p.add_argument("--brief", default="sanctuary·overworld")
    p.add_argument("--tick", type=int, default=0)
    p.add_argument("--rate", type=float, default=1.0)
    a = p.parse_args()
    asyncio.run(run(a.brief, a.tick, a.rate))


if __name__ == "__main__":
    main()

File: crates/world-core/src/gaia.rs
// gaia.rs — the world-memory: one seed, eight layers, deterministic forever.
//
// GAIA is the engine's central world-memory module. Every physical
// constant of a zone folds out of ONE pseudorandom seed: weather,
// universe entropy, time, gravity, wind, temperature, light, memory.
// Same brief + same tick ⇒ same universe.
//
// Bit-identical to quantTernEngine/gaia.ts (Node) and examples/gaia.py
// (Python) — this crate is the shared core all three surfaces link.

use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};

pub const LAYERS: [&str; 8] = [
    "time", "weather", "entropy", "gravity", "wind", "temp", "light", "memory",
];
pub const WEATHER: [&str; 8] = [
    "calm", "mist", "wind", "rain", "storm", "clear", "overcast", "frost",
];

/// The eight layers, folded from one brief at one tick.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct GaiaState {
    /// time — the zone's tick / epoch
    #[serde(rename = "t")]
    pub t: u64,
    /// weather — a word from the zone's own vocabulary
    #[serde(rename = "w")]
    pub weather: String,
    /// entropy — monotone, never rewinds
    #[serde(rename = "e")]
    pub entropy: f64,
    /// gravity — the zone's constant field
    #[serde(rename = "g")]
    pub gravity: f64,
    /// wind — direction, speed, gust
    #[serde(rename = "v")]
    pub wind: [f64; 3],
    /// temperature — the season's breath
    #[serde(rename = "p")]
    pub temp: f64,
    /// light — the sun's arc, 0..1
    #[serde(rename = "l")]
    pub light: f64,
    /// memory — the history pointer (hash chain)
    #[serde(rename = "m")]
    pub memory: String,
}

/// whash — deterministic window hash: one value per (base, epoch) pair.
/// The exact 32-bit sequence JS and Python use.
fn whash(base: u32, epoch: u32) -> u32 {
    let mut h = base ^ epoch.wrapping_mul(2654435761);
    h = (h ^ (h >> 16)).wrapping_mul(2246822519);
    h = (h ^ (h >> 13)).wrapping_mul(3266489917);
    h ^ (h >> 16)
}

/// base_of — sha256('gaia·' + brief)[:8] as a u32. The same root number
/// every implementation derives.
fn base_of(brief: &str) -> u32 {
    let mut hasher = Sha256::new();
    hasher.update(format!("gaia·{brief}").as_bytes());
    let d = hasher.finalize();
    let bytes: &[u8] = d.as_slice();
    let hex: String = bytes[..4].iter().map(|b| format!("{b:02x}")).collect();
    u32::from_str_radix(&hex, 16).expect("8 hex chars")
}

/// JS-style Math.round(x*100)/100 — half-up to two decimals.
fn r2(x: f64) -> f64 {
    (x * 100.0 + 0.5).floor() / 100.0
}

/// gaia_state — the fold: brief + tick → the eight layers.
pub fn gaia_state(brief: &str, tick: u64) -> GaiaState {
    let base = base_of(brief);
    let t = tick;

    let weather = WEATHER[(whash(base, (t / 300) as u32) as usize) % WEATHER.len()].to_string();
    let entropy = r2(t as f64 / 86400.0 + (whash(base, 0) % 1000) as f64 / 1000.0);
    let gravity = r2(9.7 + (whash(base, 1) % 100) as f64 / 100.0);

    let w0 = whash(base, (t / 60) as u32) % 360;
    let w1 = whash(base, (t / 60) as u32 + 1) % 360;
    let f = (t % 60) as f64 / 60.0;
    let dir = ((w0 as f64 + (((w1 + 360 - w0) % 360) as f64 * f)).round()) % 360.0;
    let speed = 1 + (whash(base, (t / 60) as u32 + 7) % 12);
    let gust = r2(speed as f64 + (whash(base, t as u32) % 100) as f64 / 100.0 * 4.0);
    let wind = [dir, speed as f64, gust];

    let temp = r2(15.0
        + 8.0 * ((t as f64 / 864000.0) * std::f64::consts::TAU).sin()
        + 4.0 * ((t as f64 / 86400.0) * std::f64::consts::TAU).sin()
        + ((whash(base, (t % 3600) as u32) % 100) as f64 / 100.0 - 0.5) * 2.0);
    let light = r2(((t as f64 / 86400.0) * std::f64::consts::TAU)
        .sin()
        .max(0.0));

    // memory — the history pointer: a hash chain over the folded hours
    let folds = (t / 3600) as u32;
    let mut mem = whash(base, 0);
    for i in 1..=folds.min(4096) {
        mem = whash(mem, i);
    }
    let memory = format!("{mem:08x}");

    GaiaState {
        t,
        weather,
        entropy,
        gravity,
        wind,
        temp,
        light,
        memory,
    }
}

/// gaia_wire — the compact mesh frame (single-char keys, the ternary
/// wire discipline). Field order matches gaia.ts / gaia.py.
pub fn gaia_wire(s: &GaiaState) -> String {
    serde_json::to_string(s).expect("gaia state serializes")
}

#[cfg(test)]
mod tests {
    use super::*;

    // fixtures captured from the verified Node and Python runs —
    // cross-language determinism is the contract, not a preference
    #[test]
    fn matches_node_and_python_at_t0() {
        let s = gaia_state("sanctuary·overworld", 0);
        assert_eq!(s.weather, "storm");
        assert_eq!(s.entropy, 0.37);
        assert_eq!(s.gravity, 9.74);
        assert_eq!(s.wind, [212.0, 4.0, 6.88]);
        assert_eq!(s.temp, 15.44);
        assert_eq!(s.light, 0.0);
        assert_eq!(s.memory, "499c32bc");
    }

    #[test]
    fn matches_node_and_python_at_t86400() {
        let s = gaia_state("sanctuary·overworld", 86400);
        assert_eq!(s.weather, "rain");
        assert_eq!(s.entropy, 1.37);
        assert_eq!(s.gravity, 9.74);
        assert_eq!(s.wind, [27.0, 11.0, 13.4]);
        assert_eq!(s.temp, 20.14);
        assert_eq!(s.light, 0.0);
        assert_eq!(s.memory, "167893c7");
    }

    #[test]
    fn entropy_is_monotone_and_gravity_is_constant() {
        let a = gaia_state("sanctuary·overworld", 0);
        let b = gaia_state("sanctuary·overworld", 86400);
        assert!(b.entropy > a.entropy);
        assert_eq!(a.gravity, b.gravity);
    }

    #[test]
    fn wire_is_compact() {
        let s = gaia_state("sanctuary·overworld", 0);
        let wire = gaia_wire(&s);
        assert!(wire.contains("\"t\":0"));
        assert!(wire.contains("\"w\":\"storm\""));
        assert!(wire.len() < 120, "the mesh frame stays small: {wire}");
    }
}
